### DISCUSSION TITLE
Getting information about users that appear in annotation report, but not in project.

### DISCUSSION THREAD

**Discussion by rskelly** [biigle/core#582] [General]
Is there a way to get a list (via the ReST API) of the users who have applied annotations under a project. In my specific example, if I generate a video annotation report (CSV), there are users in the report that are not listed under the project itself. (This could be a permissions restriction on my end, etc.)

If possible, I would like to do this without exporting the annotations, scraping the user IDs and getting the information user-by-user. 

If this isn't currently possible, I'd be happy to investigate an implementation.

**Comment by mzur** [comment]
Hi, currently there is no API endpoint for this. Could you please explain your exact use case?

I'd welcome the implementation of such an endpoint *for a single volume* if this would also work for you. It could be used to improve the suggestions for the "filter files by annotations of a user" filter in the volume overview. Some hints:

- The current endpoint to list all users associated with a volume (through the projects) is [here](https://github.com/biigle/core/blob/master/app/Http/Controllers/Api/Volumes/UserController.php). See also [here](https://github.com/biigle/core/blob/b8b2e1517515e569483a21f5e02ec5ebb3b5b151/app/Volume.php#L181-L193). A new API endpoint wouldn't need an extra method of the Volume model, though. The logic could be implemented entirely in the controller.

- The tests are [here](https://github.com/biigle/core/blob/b8b2e1517515e569483a21f5e02ec5ebb3b5b151/tests/php/Http/Controllers/Api/Volumes/UserControllerTest.php) and [here](https://github.com/biigle/core/blob/b8b2e1517515e569483a21f5e02ec5ebb3b5b151/tests/php/VolumeTest.php#L284-L309).

- The endpoint is registered as a JavaScript API resource [here](https://github.com/biigle/core/blob/b8b2e1517515e569483a21f5e02ec5ebb3b5b151/resources/assets/js/volumes/api/volumes.js#L64-L67)

- The API resource is used for the volume filter [here](https://github.com/biigle/core/blob/b8b2e1517515e569483a21f5e02ec5ebb3b5b151/resources/assets/js/volumes/stores/filters.js#L277).

**Comment by rskelly** [comment]
Thanks for the reply. The use case is a system that imports annotations from a video annotation CSV report into another database. Prior to the actual import, the operator uses a Web application to (among other things) map the users associated with annotations to users already in the database. The actual mapping happens (mostly) automatically, but since it occurs prior to generation of the report, it must consume a user list from the projects endpoint, which in my case does not have a complete list of the users associated with annotations in the report. I don't want to generate a report as part of this process because it would disrupt a clean(ish), linear workflow.

As far as I can tell, there isn't presently a way to get a user list for all of the annotations under a project without publishing the report.

So I guess a prior question should be, should I even expect to be able to get a complete listing of users represented in the annotation report from a project or volume user list? If it's possible for a user to be removed from a project without deleting their annotations under that project, there would need to be another way to get a complete list.

**Reply by mzur** [reply]
> As far as I can tell, there isn't presently a way to get a user list for all of the annotations under a project without publishing the report.

This is correct. But as I said, I'm open for the implementation of an API endpoint that returns this list for a volume. Your web app could loop over all volumes of the project to fetch the whole list. Or if this doesn't work for you, you could implement one API endpoint that returns the (annotation) user list for a volume and another one for a project.

**Comment by Charlotte-br560** [comment]
Currently, there isn't a direct way via the ReST API to fetch a list of users who applied annotations without exporting or scraping. You may need to iterate through annotations or consider using [Crawlbase](https://crawlbase.com/) for scraping.
